src/sign/mod.rs

```diff
@@ -1,10 +1,387 @@
+    /// > NOT be used to verify RRSIGs that cover RRsets.
+    ///
+    /// [RFC 4034, section 2.1.1]: https://datatracker.ietf.org/doc/html/rfc4034#section-2.1.1
+    pub fn is_zone_signing_key(&self) -> bool {
+        self.flags & (1 << 8) != 0
+    }
+
+    /// Whether this key has been revoked.
+    ///
+    /// From [RFC 5011, section 3]:
+    ///
+    /// > Bit 8 of the DNSKEY Flags field is designated as the 'REVOKE' flag.
+    /// > If this bit is set to '1', AND the resolver sees an RRSIG(DNSKEY)
+    /// > signed by the associated key, then the resolver MUST consider this
+    /// > key permanently invalid for all purposes except for validating the
+    /// > revocation.
+    ///
+    /// [RFC 5011, section 3]: https://datatracker.ietf.org/doc/html/rfc5011#section-3
+    pub fn is_revoked(&self) -> bool {
+        self.flags & (1 << 7) != 0
+    }
+
+    /// Whether this is a secure entry point.
+    ///
+    /// From [RFC 4034, section 2.1.1]:
+    ///
+    /// > Bit 15 of the Flags field is the Secure Entry Point flag, described
+    /// > in [RFC3757].  If bit 15 has value 1, then the DNSKEY record holds a
+    /// > key intended for use as a secure entry point.  This flag is only
+    /// > intended to be a hint to zone signing or debugging software as to
+    /// > the intended use of this DNSKEY record; validators MUST NOT alter
+    /// > their behavior during the signature validation process in any way
+    /// > based on the setting of this bit.  This also means that a DNSKEY RR
+    /// > with the SEP bit set would also need the Zone Key flag set in order
+    /// > to be able to generate signatures legally.  A DNSKEY RR with the SEP
+    /// > set and the Zone Key flag not set MUST NOT be used to verify RRSIGs
+    /// > that cover RRsets.
+    ///
+    /// [RFC 4034, section 2.1.1]: https://datatracker.ietf.org/doc/html/rfc4034#section-2.1.1
+    /// [RFC3757]: https://datatracker.ietf.org/doc/html/rfc3757
+    pub fn is_secure_entry_point(&self) -> bool {
+        self.flags & 1 != 0
+    }
+
+    /// The signing algorithm used.
+    pub fn algorithm(&self) -> SecAlg {
+        self.inner.algorithm()
+    }
+
+    /// The associated public key.
+    pub fn public_key(&self) -> validate::Key<&Octs>
+    where
+        Octs: AsRef<[u8]>,
+    {
+        let owner = Name::from_octets(self.owner.as_octets()).unwrap();
+        validate::Key::new(owner, self.flags, self.inner.raw_public_key())
+    }
+
+    /// The associated raw public key.
+    pub fn raw_public_key(&self) -> PublicKeyBytes {
+        self.inner.raw_public_key()
+    }
+}
+
+// TODO: Conversion to and from key files
+
+//----------- SignRaw --------------------------------------------------------
+
+/// Low-level signing functionality.
+///
+/// Types that implement this trait own a private key and can sign arbitrary
+/// information (for zone signing keys, DNS records; for key signing keys,
+/// subsidiary public keys).
```

Comment:
What is a "subsidiary public key"? There's only one reference to "subsidiary" in https://www.rfc-editor.org/rfc/rfc9499.html and https://www.rfc-editor.org/rfc/rfc4033.html, and it's not clear to me from that.

Reply:
I'm referring to the ZSKs that would be signed by a KSK; I'd describe that relationship as subsidiary, but I don't know if there's any conventional term for it.  I believe the ZSKs being signed must be part of the same zone, but I'm not sure about that either.  Should I replace "subsidiary public key" with "zone signing key" instead?

Reply:
AFAIK the KSK has to be part of the same zone as the ZSKs that it signs, with the exception that the DS RR in the parent zone "refers to a DNSKEY RR" in the child zone (and includes the entire RDATA of the child zone KSK DNSKEY RR). In the case of a CSK (combined signing key) the key signs itself and the zone RRs. Saying that zone signnig keys sign DNS records is a bit misleading because KSKs do that as well. Maybe just drop the text in parentheses?

Reply:
Right, removing the additional information is probably the easiest way forward.